### DISCUSSION TITLE
Where is type defined and: Is it a vectorized word? rp2040_big

### DISCUSSION THREAD

**Discussion by ghost** [tabemann/zeptoforth#155] [General]
I've searched all files in src but couldn't find the definition of type. Only some "type" (c-code?) for other flavors as rp2040.
IMHO type should be vectorized but if I try the following zeptoforth crashes.
' type defer@ ' type defer!
Tanks in advance!

**Comment by tabemann** [comment]
On Fri, Nov 8, 2024 at 6:07 AM MB ***@***.***> wrote:

> I've searched all files in src but couldn't find the definition of type.
> Only some "type" (c-code?) for other flavors as rp2040.
> IMHO type should be vectorized but if I try the following zeptoforth
> crashes.
> ' type defer@ ' type defer!
> Tanks in advance!
>
For starters, TYPE is not an DEFER-ed word, so that won't work -- it's
guaranteed to give you undefined behavior. Also, the words that actually
control console I/O are EMIT , EMIT? , KEY , KEY? , and FLUSH-CONSOLE whose
redirection is set for any given task through the task-local variables
EMIT-HOOK , EMIT?-HOOK , KEY-HOOK , KEY?-HOOK , and FLUSH-CONSOLE-HOOK
(there are also ERROR-EMIT-HOOK and ERROR-EMIT?-HOOK for setting the
redirection of error output). For more information look in
src/common/forth/console.fs . (About where TYPE itself is defined, it is
defined in src/m0/core.s , src/m4_m7/core.s , and src/m33/core.s ─ note
that this code is in assembly, not in Forth.)

Travis

**Comment by tabemann** [comment]
On Fri, Nov 8, 2024 at 9:34 AM Travis Bemann ***@***.***> wrote:

> On Fri, Nov 8, 2024 at 6:07 AM MB ***@***.***> wrote:
>
>> I've searched all files in src but couldn't find the definition of type.
>> Only some "type" (c-code?) for other flavors as rp2040.
>> IMHO type should be vectorized but if I try the following zeptoforth
>> crashes.
>> ' type defer@ ' type defer!
>> Tanks in advance!
>>
> For starters, TYPE is not an DEFER-ed word, so that won't work -- it's
> guaranteed to give you undefined behavior. Also, the words that actually
> control console I/O are EMIT , EMIT? , KEY , KEY? , and FLUSH-CONSOLE whose
> redirection is set for any given task through the task-local variables
> EMIT-HOOK , EMIT?-HOOK , KEY-HOOK , KEY?-HOOK , and FLUSH-CONSOLE-HOOK
> (there are also ERROR-EMIT-HOOK and ERROR-EMIT?-HOOK for setting the
> redirection of error output). For more information look in
> src/common/forth/console.fs . (About where TYPE itself is defined, it is
> defined in src/m0/core.s , src/m4_m7/core.s , and src/m33/core.s ─ note
> that this code is in assembly, not in Forth.)
>

I should also note that DEFER-ed words in flash can only be DEFER!-ed/IS-ed
once (and strictly speaking this is not guaranteed to work if the system
has rebooted in between when the DEFER-ed word was originally created and
when it was set), so for general vectorization purposes it is preferable to
create HOOK variables for vectorization, as in the following:

variable foo-hook
: foo foo-hook @ ?execute ;
: init-foo 0 foo-hook ! ;
initializer init-foo

Travis
